### PR TITLE
NSLock: Implement lock(before:)

### PR DESCRIPTION
**Pull request by spevans**
- Implemented for NSLock and NSRecursiveLock.

- Use pthread_mutex_timedlock() where available (eg Linux).

- For macOS / iOS, emulate a timed lock using a condition
  variable/mutex pair and use pthread_cond_timedwait().

- Add timeSpecFrom(date:) helper function and update
  NSCondition.wait(until:) to use it, eliminating two calls
  to gettimeofday().

The testing just consists of creating 1000 threads and then have each thread trying to get the lock in a loop along with some other contrived lock activity. Some more testing will probably need to be added to try and force some errors and also stress the thread/lock system.

`getOrAllocateTimedLockData()` needs the most attention as that uses `_stdlib_atomicCompareExchangeStrongPtr` to atomically set the shared mutex/cond pair used on macOS and the use of allocated pointers etc could be incorrect.

### CODE REVIEW

File: TestFoundation/TestNSLock.swift [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        ]
+    }
+
+
+    private func debugLog(_ msg: String) {
+#if false
+        NSLog(msg)
+#endif
+    }
+
```
Comment (parkera) [inline TestFoundation/TestNSLock.swift]: Nit pick, but let's just leave out this kind of temporary debug stuff.

File: TestFoundation/TestNSLock.swift [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        } while totalThreads < threadCount
+        XCTAssertEqual(totalThreads, threadCount)
+
+        let gotLock = lock.try()
+        XCTAssertTrue(gotLock)
+        if gotLock {
+            lock.unlock()
+        }
+    }
+}
```
Comment (parkera) [inline TestFoundation/TestNSLock.swift]: Out of curiosity, how long do these tests take to run? Is there going to be a problem with creating so many threads on some platforms?

Comment (spevans) [reply TestFoundation/TestNSLock.swift]: I just ran the tests on my 2014 Macbook Pro and they took 8.008 and 10.003 seconds so about 18seconds for both tests.

1000 threads were created in .15secs. I was originally creating 10,000 but the limit looks to be 2048 total (on macOS)  and I could only create 2044. It didn't error out but I think `Thread.start` could be modified to get the error result from `pthread_create` and set the status appropriately.

I guess some low end devices will have a problem but the `threadCount` can just be lowered, the idea is really to try and stress the system to force some locking errors.

I have some more tests to add at some point for the `NSConditionLock` so I'll tidy up the debug then.

**Comment by ianpartridge** [comment]
Nice. I didn't know about the stdlib intrinsics for cmpxchg.

**Comment by ianpartridge** [comment]
@swift-ci please test

File: Foundation/NSLock.swift [hunk 3]
```diff
@@ -36,24 +43,47 @@ open class NSLock: NSObject, NSLocking {
...
     open func lock() {
         pthread_mutex_lock(mutex)
     }
-    
+
     open func unlock() {
         pthread_mutex_unlock(mutex)
+#if os(OSX) || os(iOS)
+        if let tld = timedLockData {
+            pthread_mutex_lock(&tld.pointee.mutex)
```
Comment (phausler) [inline Foundation/NSLock.swift]: I am not fully certain this is safe since swift can potentially write into the mutex memory when it takes the address of the mutex as a stack based copy

File: Foundation/NSLock.swift [hunk 4]
```diff
@@ -258,3 +286,80 @@ open class NSCondition: NSObject, NSLocking {
...
+private func getOrAllocateTimedLockData(timedLockData: inout UnsafeMutablePointer<_timed_lock_data>?) -> UnsafeMutablePointer<_timed_lock_data> {
+    if timedLockData == nil, let tld = allocateTimedLockData() {
+        var expected: UnsafeMutablePointer<_timed_lock_data>? = nil
+
+        // Use an atomic compare and exchange to set the pointer to the just allocated
+        // data. On X86 this is a LOCK CMPXCHG.
+        // `object` is the target to update, timedLockData, shared by all threads.
+        // `expected` is the value the target should still be (nil).
+        // `desired` is the new value to set it to.
+        if !_stdlib_atomicCompareExchangeStrongPtr(object: &timedLockData,
```
Comment (phausler) [inline Foundation/NSLock.swift]: oof! are we certain that taking the address here is safe? is it actually atomic?

using _ prefix methods from the standard library should always be done with their consent imho since we are effectively making it API.

@airspeedswift What is your take on using this?

File: Foundation/NSLock.swift [hunk 5]
```diff
@@ -258,3 +286,80 @@ open class NSCondition: NSObject, NSLocking {
...
+
+// Emulate pthread_mutex_timedlock using pthread_cond_timedwait.
+// lock(before:) has obtained a shared pointer to a _time_lock_data structure
+// and the condition variable to be waited on.
+// unlock() will check for the existance of this structure and if its pointer is
+// non-nil, it will use pthread_cond_broadcast() to stop any waits in progress.
+private func timedLock(mutex: _PthreadMutexPointer,
+                       endTime: inout timespec,
+                       using tld: UnsafeMutablePointer<_timed_lock_data>) -> Bool {
+    pthread_mutex_lock(&tld.pointee.mutex)
```
Comment (phausler) [inline Foundation/NSLock.swift]: again I am not certain this is safe

File: Foundation/NSLock.swift (hunk 4 above)
Comment (spevans) [reply Foundation/NSLock.swift]: I checked the disassembly (on x86_64) and it looked correct. My biggest concern was actually that the documentation seems to imply that the `expect` pointer should be heap allocated but I wasn't sure if that was correct.

I understand your point about _ prefix methods, however we use `@_` function decorators, is that not the same violation? Getting a definitive answer from the Swift core team would be good on both of these points.

I guess as an alternative this could always be done in C in CoreFoundation instead, bypassing these issues.

Comment (phausler) [reply Foundation/NSLock.swift]: for the specific cases we use `@_` either I have been chastised for using that feature and eventually have been forced to remove it or we have been given special dispensation to utilize it usually with the caveat that when they change things we must change as well.

Comment (airspeedswift) [reply Foundation/NSLock.swift]: We should definitely not be calling `_stdlib_atomicCompareExchangeStrongPtr` from std lib code.

Comment (airspeedswift) [reply Foundation/NSLock.swift]: Checking the disassembly is only guaranteeing that it's correct _now_. It could easily change and be incorrect later if it's relying on undefined behavior. But I think (assuming the _ method use is removed) that this usage is correct so long as it is using allocated memory not a property for the `pointee`. cc @jckarter

Comment (airspeedswift) [reply Foundation/NSLock.swift]: Oops that should read "We should definitely not be calling `_stdlib_atomicCompareExchangeStrongPtr` from *outside* the std lib."

Comment (spevans) [reply Foundation/NSLock.swift]: @airspeedswift What would you recommend as an alternative?

Comment (spevans) [reply Foundation/NSLock.swift]: I changed the `_stdlib_atomicCompareExchangeStrongPtr` to using a `lazy var`

File: Foundation/NSLock.swift (hunk 5 above)
Comment (spevans) [reply Foundation/NSLock.swift]: @phausler Who would be able to give a definitive answer as to whether it is safe or not?

Comment (jckarter) [reply Foundation/NSLock.swift]: Formally it isn't guaranteed to be safe, but I can't think of anything better today.

Comment (parkera) [reply Foundation/NSLock.swift]: I think we'll need to implement it in C and call out from Swift.

Comment (parkera) [reply Foundation/NSLock.swift]: The performance may not be great, but that is a secondary concern to correctness.

Comment (spevans) [reply Foundation/NSLock.swift]: At least doing it in C saves adding a 2nd ptr to every `Lock` allocation which will mostly be wasted (except for the locks using the `lock(before:)` method. I will look into it.

Comment (jckarter) [reply Foundation/NSLock.swift]: A `static inline` C function to project the pointer to the mutex field from a pointer to the struct could also still be inlined away.

**Comment by spevans** [comment]
@swift-ci please test

**Comment by ianpartridge** [comment]
@swift-ci please test

**Comment by ianpartridge** [comment]
@swift-ci please test

**Comment by spevans** [comment]
@parkera I think this is ready to merge now

**Comment by spevans** [comment]
@swift-ci please test

**Comment by xwu** [comment]
@spevans Drive-by commentary: lazy var isn't thread-safe; did you change the use of _stdlib_atomicCompareExchangeStrongPtr to lazy var on the assumption that it is?

File: CoreFoundation/Base.subproj/ForSwiftFoundationOnly.h [hunk 6]
```diff
@@ -307,6 +307,21 @@ CF_EXPORT _CFThreadSpecificKey _CFThreadSpecificKeyCreate(void);
...
 typedef pthread_t _CFThreadRef;
 
+// _timed_lock_data is a condition variable and mutex pair used to wait for
+// a lock to be unlocked by another thread. See NSLock lock(before:).
+struct _timed_lock_data {
+    pthread_mutex_t mutex;
+    pthread_cond_t cond;
+};
+
+CF_INLINE pthread_mutex_t *_Nonnull _CFGetTimedLockMutex(struct _timed_lock_data *_Nonnull data) {
```
Comment (weissi) [inline CoreFoundation/Base.subproj/ForSwiftFoundationOnly.h]: why are these needed? Isn't `&timedLockData.mutex` all you need?

File: Foundation/NSLock.swift [hunk 7]
```diff
@@ -36,24 +47,51 @@ open class NSLock: NSObject, NSLocking {
...
     }
-    
+
     open func unlock() {
         pthread_mutex_unlock(mutex)
+#if os(OSX) || os(iOS)
+        // Only use _timedLockData if it was initialized by lock(before:)
+        if let tld = _timedLockData {
+            pthread_mutex_lock(_CFGetTimedLockMutex(tld))
+            pthread_cond_broadcast(_CFGetTimedLockCond(tld))
```
Comment (weissi) [inline Foundation/NSLock.swift]: - why do we use `pthread_cond_broadcast` instead of `pthread_cond_signal`
 - all the return values aren't checked (it's fine to `precondition(ret == 0)`)

File: Foundation/NSLock.swift [hunk 8]
```diff
@@ -143,12 +181,15 @@ open class NSConditionLock : NSObject, NSLocking {
...
 open class NSRecursiveLock: NSObject, NSLocking {
-#if CYGWIN
-    internal var mutex = UnsafeMutablePointer<pthread_mutex_t?>.allocate(capacity: 1)
-#else
-    internal var mutex = UnsafeMutablePointer<pthread_mutex_t>.allocate(capacity: 1)
+    internal var mutex = _PthreadMutexPointer.allocate(capacity: 1)
+#if os(OSX) || os(iOS)
+    private var _timedLockData: UnsafeMutablePointer<_timed_lock_data>? = nil
+    private lazy var timedLockData: UnsafeMutablePointer<_timed_lock_data>? = {
+        _timedLockData = allocateTimedLockData()
```
Comment (weissi) [inline Foundation/NSLock.swift]: not thread safe

File: Foundation/NSLock.swift [hunk 9]
```diff
@@ -174,28 +218,47 @@ open class NSRecursiveLock: NSObject, NSLocking {
...
     
-    open func lock(before limit: Date) {
-        NSUnimplemented()
+    open func lock(before limit: Date) -> Bool {
+        if pthread_mutex_trylock(mutex) == 0 {
+            return true
+        }
+
+        guard var endTime = timeSpecFrom(date: limit) else {
+            return false
```
Comment (weissi) [inline Foundation/NSLock.swift]: why do we return `false` if `timeSpecFrom(date:)` fails? It should crash here, `false` means the timeout expired.

File: Foundation/NSLock.swift [hunk 10]
```diff
@@ -174,28 +218,47 @@ open class NSRecursiveLock: NSObject, NSLocking {
...
+        }
+
+#if os(OSX) || os(iOS)
+        if let tld = timedLockData {
+            return timedLock(mutex: mutex, endTime: &endTime, using: tld)
+        } else {
+            return false
+        }
+#else
+        return pthread_mutex_timedlock(mutex, &endTime) == 0
```
Comment (weissi) [inline Foundation/NSLock.swift]: that puts all errors in the same category. Should be

```
            switch pthread_mutex_timedlock(mutex, &endTime) {
            case 0:
                return true
            case ETIMEDOUT:
                return false
            case let e:
                fatalError("caught error \(e) when calling pthread_cond_timedwait")
            }
```

File: Foundation/NSLock.swift [hunk 11]
```diff
@@ -221,31 +284,12 @@ open class NSCondition: NSObject, NSLocking {
...
-                ts.tv_sec += ts.tv_nsec / 1_000_000_000
-                ts.tv_nsec = ts.tv_nsec % 1_000_000_000
-            }
-        }
-        let retVal: Int32 = withUnsafePointer(to: &ts) { t in
-            return pthread_cond_timedwait(cond, mutex, t)
-        }
-
-        return retVal == 0
+        return pthread_cond_timedwait(cond, mutex, &timeout) == 0
```
Comment (weissi) [inline Foundation/NSLock.swift]: same here, errors besides `ETIMEDOUT` should be handled separately.

File: Foundation/NSLock.swift [hunk 12]
```diff
@@ -258,3 +302,59 @@ open class NSCondition: NSObject, NSLocking {
...
+
+private func timeSpecFrom(date: Date) -> timespec? {
+    let interval = date.timeIntervalSince1970
+    guard interval >= 0 else {
+        return nil
+    }
+
+    var ts = timespec()
+    ts.tv_sec = Int(interval.rounded(.down))
+    ts.tv_nsec = Int((interval - Double(ts.tv_sec)) * 1_000_000_000.0)
```
Comment (weissi) [inline Foundation/NSLock.swift]: this looks rather dangerous. If for floating point reasons `ts.tv_nsec` will be `1_000_000_000` or larger, all the timed wait functions will just return `EINVAL` IIRC.

Why not something similar to

```
private func timeSpecFrom(date: Date) -> timespec {
        guard interval >= 0 else { fatalError ("current time negative") }
        let nsecPerSec: Int64 = 1000000000
        let timeoutNS = Int64(interval * Double(nsecPerSec))

        var timeoutAbs = timespec(tv_sec: Int((timeoutNS / nsecPerSec)),
                                  tv_nsec: Int(timeoutNS % nsecPerSec))
        assert(timeoutAbs.tv_nsec >= 0 && timeoutAbs.tv_nsec < Int(nsecPerSec))
        assert(timeoutAbs.tv_sec >= curTime.tv_sec)
        return timeoutAbs
}
```

File: Foundation/NSLock.swift [hunk 13]
```diff
@@ -258,3 +302,59 @@ open class NSCondition: NSObject, NSLocking {
...
+
+#if os(OSX) || os(iOS)
+
+private func allocateTimedLockData() -> UnsafeMutablePointer<_timed_lock_data>? {
+    let tld = UnsafeMutablePointer<_timed_lock_data>.allocate(capacity: 1)
+    if pthread_mutex_init(_CFGetTimedLockMutex(tld), nil) == 0 {
+        if pthread_cond_init(_CFGetTimedLockCond(tld), nil) == 0 {
+            return tld
+        } else {
+            pthread_mutex_destroy(_CFGetTimedLockMutex(tld))
```
Comment (weissi) [inline Foundation/NSLock.swift]: need to check the error code

File: Foundation/NSLock.swift [hunk 14]
```diff
@@ -258,3 +302,59 @@ open class NSCondition: NSObject, NSLocking {
...
+
+    var ts = timespec()
+    ts.tv_sec = Int(interval.rounded(.down))
+    ts.tv_nsec = Int((interval - Double(ts.tv_sec)) * 1_000_000_000.0)
+    return ts
+}
+
+#if os(OSX) || os(iOS)
+
+private func allocateTimedLockData() -> UnsafeMutablePointer<_timed_lock_data>? {
```
Comment (weissi) [inline Foundation/NSLock.swift]: in Swift, failure to allocate memory usually results in a crash. Why does this return an Optional instead? What's the caller supposed to do if this fails?

File: Foundation/NSLock.swift [hunk 15]
```diff
@@ -258,3 +302,59 @@ open class NSCondition: NSObject, NSLocking {
...
+            return tld
+        } else {
+            pthread_mutex_destroy(_CFGetTimedLockMutex(tld))
+        }
+    }
+    tld.deallocate(capacity: 1)
+    return nil
+}
+
+private func deallocateTimedLockData(_ tld: UnsafeMutablePointer<_timed_lock_data>?) {
```
Comment (weissi) [inline Foundation/NSLock.swift]: shouldn't take an optional

File: Foundation/NSLock.swift [hunk 16]
```diff
@@ -258,3 +302,59 @@ open class NSCondition: NSObject, NSLocking {
...
+
+// Emulate pthread_mutex_timedlock using pthread_cond_timedwait.
+// lock(before:) has obtained a shared pointer to a _time_lock_data structure
+// and the condition variable to be waited on.
+// unlock() will check for the existance of this structure and if its pointer is
+// non-nil, it will use pthread_cond_broadcast() to stop any waits in progress.
+private func timedLock(mutex: _PthreadMutexPointer,
+                       endTime: inout timespec,
+                       using tld: UnsafeMutablePointer<_timed_lock_data>) -> Bool {
+    pthread_mutex_lock(_CFGetTimedLockMutex(tld))
```
Comment (weissi) [inline Foundation/NSLock.swift]: errors not checked

File: Foundation/NSLock.swift [hunk 17]
```diff
@@ -258,3 +302,59 @@ open class NSCondition: NSObject, NSLocking {
...
+    let status = pthread_cond_timedwait(_CFGetTimedLockCond(tld),
+                                        _CFGetTimedLockMutex(tld),
+                                        &endTime)
+    pthread_mutex_unlock(_CFGetTimedLockMutex(tld))
```
Comment (weissi) [inline Foundation/NSLock.swift]: wait, why are we unlocking here?

File: Foundation/NSLock.swift [hunk 18]
```diff
@@ -258,3 +302,59 @@ open class NSCondition: NSObject, NSLocking {
...
+
+    if status == 0 {
+        return pthread_mutex_trylock(mutex) == 0
```
Comment (weissi) [inline Foundation/NSLock.swift]: why are we trying to re-lock here?

File: Foundation/NSLock.swift (hunk 9 above)
Comment (spevans) [reply Foundation/NSLock.swift]: Crashing is a little extreme because a timeout was exceeded. The documentation at https://developer.apple.com/documentation/foundation/nslock/1411133-lock states that the return value is `true if the lock is acquired before limit, otherwise false.`
The `lock(before:)` could have been called with a perfectly valid date in the future but could have expired before it is processed (e.g. if it was something in the near future like 3ns)

File: Foundation/NSLock.swift (hunk 15 above)
Comment (spevans) [reply Foundation/NSLock.swift]: This takes an optional to simplify the `deinit`. Its a `private` function so its not exposing any sort of api.

File: Foundation/NSLock.swift (hunk 14 above)
Comment (spevans) [reply Foundation/NSLock.swift]: This was copied from the `-[NSLock lock(before:)]` function but you are correct, it should be stricter with the checking, I will update it to call `fatalError()` if anything fails.

File: Foundation/NSLock.swift (hunk 17 above)
Comment (spevans) [reply Foundation/NSLock.swift]: `pthread_cond_timedwait` requires the lock so it needs to be unlocked.

File: Foundation/NSLock.swift (hunk 16 above)
Comment (phausler) [reply Foundation/NSLock.swift]: if pthread_mutex_lock fails I would consider the host OS to potentially be defective; The errors are EINVAL - which means the lock is corrupt, or EDEADLK which only happens when the lock is initialized with the parameters PTHREAD_MUTEX_ERRORCHECK

File: Foundation/NSLock.swift (hunk 13 above)
Comment (phausler) [reply Foundation/NSLock.swift]: not needed; the failure cases are EINVAL for a corrupted lock, or EBUSY which only happens in the cases of PTHREAD_MUTEX_ERRORCHECK

File: Foundation/NSLock.swift (hunk 7 above)
Comment (phausler) [reply Foundation/NSLock.swift]: there may be multiple waiters on a given NSLock so broadcast is the correct idiom

File: Foundation/NSLock.swift (hunk 18 above)
Comment (spevans) [reply Foundation/NSLock.swift]: This is how the lock is actually acquired. The `pthread_mutex_lock` / `pthread_cond_timedwait` / `pthread_mutex_unlock` merely handles the waiting/timeout, the `pthread_mutex_trylock` is the actual (non-waiting) lock.

File: Foundation/NSLock.swift [hunk 19]
```diff
@@ -21,13 +21,24 @@ public protocol NSLocking {
...
+private typealias _PthreadMutexPointer = UnsafeMutablePointer<pthread_mutex_t>
+private typealias _PthreadCondPointer = UnsafeMutablePointer<pthread_cond_t>
 #endif
-    
+
+open class NSLock: NSObject, NSLocking {
+    internal var mutex = _PthreadMutexPointer.allocate(capacity: 1)
+#if os(OSX) || os(iOS)
+    private var _timedLockData: UnsafeMutablePointer<_timed_lock_data>? = nil
+    private lazy var timedLockData: UnsafeMutablePointer<_timed_lock_data>? = {
```
[the remark this answers is not in the document]
Comment (spevans) [reply Foundation/NSLock.swift]: What is the correct way of doing an implicit `swift_once()` with `var`s? Or should I use an explicit `dispatch_once()`?

File: Foundation/NSLock.swift (hunk 7 above)
Comment (weissi) [reply Foundation/NSLock.swift]: @phausler please correct me if I'm wrong but AFAIK _any_ of the waiters will be able to make progress and also after every unlock we run that code. So why do we need to wake up _all_ and not just one?

File: Foundation/NSLock.swift (hunk 13 above)
Comment (weissi) [reply Foundation/NSLock.swift]: but we want to know this, right? What if we have memory corruption or something. We should just `preconditon()` or at the very least `assert()` that the return value is 0.

File: Foundation/NSLock.swift (hunk 16 above)
Comment (weissi) [reply Foundation/NSLock.swift]: so why not just `precondition` or `assert`? Just to be sure.

File: Foundation/NSLock.swift (hunk 17 above)
Comment (weissi) [reply Foundation/NSLock.swift]: @spevans sorry I think I'm slow. So I agree that `pthread_cond_timedwait` needs the lock to be locked which is the case. But the function is called `timedLock` so when the function returns, the lock should still be locked, right? Otherwise the function returns and the lock isn't actually locked.

File: Foundation/NSLock.swift (hunk 18 above)
Comment (weissi) [reply Foundation/NSLock.swift]: @spevans sorry, now I'm completely lost, what is a non-waiting lock? And why do we need two locks for a timed lock?

File: Foundation/NSLock.swift (hunk 9 above)
Comment (weissi) [reply Foundation/NSLock.swift]: @spevans sorry, my comment is totally bogus. `timeSpecFrom(date:)` doesn't return `false` if the timeout expired but if it couldn't convert the timeout. That should not be possible. The right fix is obvs to make `timeSpecFrom(date:)` not even return a `Bool` as it should always succeed. Then we also don't need this `return` here.

File: Foundation/NSLock.swift (hunk 15 above)
Comment (weissi) [reply Foundation/NSLock.swift]: fair enough

File: Foundation/NSLock.swift (hunk 18 above)
Comment (spevans) [reply Foundation/NSLock.swift]: On macOS there are only two ways to lock: `pthread_mutex_lock` which waits indefinitely and `pthread_mutex_trylock` which tries to lock and returns immediately whether it locked or not (i.e. non-waiting).

`pthread_cond_timewait` uses a `pthread_mutex_t`/`pthread_cond_t` pair to perform its actions. So there are actually 2 mutexes and one condition variable involved here.

`mutex` is the actual lock that is trying to be acquired. The `_timed_lock_data` struct holds the `pthread_mutex_t`/`pthread_cond_t` pair for the `pthread_cond_timewait`.
Does that make sense?

Comment (weissi) [reply Foundation/NSLock.swift]: @spevans apologies, I didn't quite read the code correctly, thought this is all the same mutex :P.

Ok, in that case I still think it's not quite right as the `pthread_mutex_unlock(_CFGetTimedLockMutex(tld))` is done too early. If I'm not mistaken what could happen is the following: 

- thread 1 progresses until line  354 with `status == 0`
- thread 1 is now scheduled off CPU
- thread 2 arrives and also progresses until line 354 with `status == 0`
- thread 2 runs the function to completion and gets the lock `mutex` (with `pthread_mutex_trylock(mutex) == 0`)
- now thread 1 gets scheduled again and also tries to get the lock (with `pthread_mutex_trylock(mutex) == 0`) but because it's already taken by thread 2, it doesn't get it

now, in thread 1 `timedLock(...)` returns `false` even though the timeout might not have expired.


The same problem can happen if `pthread_cond_timedwait` spuriously wakes up. It's allowed to do that. Basically conditional variables are hardly ever used correctly if they're not in a loop. See [here](https://linux.die.net/man/3/pthread_cond_wait)

> Spurious wakeups from the pthread_cond_timedwait() or pthread_cond_wait() functions may occur.

please let me  know if I missed anything again and sorry agaiin

File: Foundation/NSLock.swift (hunk 19 above)
[the remark this answers is not in the document]
Comment (jckarter) [reply Foundation/NSLock.swift]: The `once` APIs only work with global variables. You'll need some other form of synchronization, like a lock.

File: Foundation/NSLock.swift (hunk 18 above)
Comment (spevans) [reply Foundation/NSLock.swift]: @weissi Yes you are correct in that I have made a small error assuming that as soon as one thread gets the lock the others should return `false`, when they should actually retry the loop until the `limit: Date` is reached. Adding a loop would fix this and also work around any spurious wake ups.

I will add a fix for the above, however note that this solution is only used on macOS and the manual page for `pthread_cond_timedwait` doesn't actually mention spurious wakeups, it will only return on an invalid argument or a timeout.

On Linux, all of this is replaced by a single call of `pthread_mutex_timedlock`.

**Comment by spevans** [comment]
@swift-ci please test

**Comment by spevans** [comment]
@swift-ci please test

**Comment by spevans** [comment]
I updated this to simplify the cond variable/mutex pair to always be allocated. Its makes the `unlock` slightly less efficient on macOS since the `pthread_cond_broadcast` always happens now but it should fix any thread safety issues due to the lazy allocation. A more better solution can alway be used in a follow up PR.

@weissi I added a loop around the `pthread_cond_timedwait` to check that the end time is reached which should solve any issues of early wakeup. I also modified the `timeSpecFrom(date:)` to use integers as I also saw a potential issue with the use of `Double` overflowing the `tv_nsec` field.

File: Foundation/NSLock.swift [hunk 20]
```diff
@@ -258,3 +286,48 @@ open class NSCondition: NSObject, NSLocking {
...
+// lock(before:) passes a condition variable/mutex pair to use.
+// unlock() will use pthread_cond_broadcast() to wake any waits in progress.
+private func timedLock(mutex: _PthreadMutexPointer, endTime: Date,
+                       using timeoutCond: _PthreadCondPointer,
+                       with timeoutMutex: _PthreadMutexPointer) -> Bool {
+
+    var timeSpec = timeSpecFrom(date: endTime)
+    while var ts = timeSpec {
+        pthread_mutex_lock(timeoutMutex)
+        pthread_cond_timedwait(timeoutCond, timeoutMutex, &ts)
```
Comment (weissi) [inline Foundation/NSLock.swift]: please, can we `precondition` that the return value of `pthread_cond_timedwait` is either `0` or `ETIMEDOUT` and nothing else.

Also we should `precondition` (at least `assert`) that `pthread_mutex_(un)lock` return `0` in every case.

File: Foundation/NSLock.swift [hunk 21]
```diff
@@ -21,39 +21,71 @@ public protocol NSLocking {
...
-    internal var mutex = UnsafeMutablePointer<pthread_mutex_t>.allocate(capacity: 1)
+private typealias _PthreadMutexPointer = UnsafeMutablePointer<pthread_mutex_t>
+private typealias _PthreadCondPointer = UnsafeMutablePointer<pthread_cond_t>
 #endif
-    
+
+open class NSLock: NSObject, NSLocking {
+    internal var mutex = _PthreadMutexPointer.allocate(capacity: 1)
+#if os(OSX) || os(iOS)
+    private var timeoutCond = _PthreadCondPointer.allocate(capacity: 1)
```
Comment (weissi) [inline Foundation/NSLock.swift]: wouldn't it make sense to unify this code and use the more portable BSD/Darwin code for all platforms?

File: Foundation/NSLock.swift [hunk 22]
```diff
@@ -21,39 +21,71 @@ public protocol NSLocking {
...
+    private var timeoutMutex = _PthreadMutexPointer.allocate(capacity: 1)
+#endif
+
     public override init() {
         pthread_mutex_init(mutex, nil)
+#if os(OSX) || os(iOS)
+        pthread_cond_init(timeoutCond, nil)
```
Comment (weissi) [inline Foundation/NSLock.swift]: why aren't we checking the return values of `pthread_*` anywhere really?

**Comment by spevans** [comment]
@swift-ci please test

**Comment by alblue** [comment]
@swift-ci please test and merge
